runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/ClockSkewInterceptor.kt

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.http.interceptors
+
+import aws.smithy.kotlin.runtime.client.ProtocolRequestInterceptorContext
+import aws.smithy.kotlin.runtime.client.ProtocolResponseInterceptorContext
+import aws.smithy.kotlin.runtime.http.operation.HttpOperationContext
+import aws.smithy.kotlin.runtime.http.request.HttpRequest
+import aws.smithy.kotlin.runtime.http.response.HttpResponse
+import aws.smithy.kotlin.runtime.http.response.header
+import aws.smithy.kotlin.runtime.telemetry.logging.logger
+import aws.smithy.kotlin.runtime.time.Instant
+import aws.smithy.kotlin.runtime.time.until
+import kotlinx.atomicfu.*
+import kotlin.coroutines.coroutineContext
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.minutes
+
+/**
+ * An interceptor used to detect clock skew (difference between client and server clocks) and apply a correction.
+ */
+public class ClockSkewInterceptor : HttpInterceptor {
+    public companion object {
+        /**
+         * How much must the clock be skewed before attempting correction
+         */
+        public val CLOCK_SKEW_THRESHOLD: Duration = 4.minutes
+
+        /**
+         * Get the current skew between the client time and [serverTime] as a [Duration].
+         * It may be negative if the serverTime is in the past.
+         * @param serverTime the server's time
+         */
+        internal fun Instant.getSkew(serverTime: Instant): Duration = this.until(serverTime)
+
+        /**
+         * Determine whether the client's clock is skewed relative to the server.
+         * @param serverTime the server's time
+         */
+        internal fun Instant.isSkewed(serverTime: Instant): Boolean = getSkew(serverTime).absoluteValue >= CLOCK_SKEW_THRESHOLD
+    }
+
+    // Clock skew to be applied to all requests
+    private val _currentSkew: AtomicRef<Duration> = atomic(Duration.ZERO)
+
+    private val currentSkew: Duration
+        get() = _currentSkew.value
+
+    /**
+     * Apply the previously-computed skew, if it's set, to the execution context before signing
+     */
+    public override suspend fun modifyBeforeSigning(context: ProtocolRequestInterceptorContext<Any, HttpRequest>): HttpRequest {
+        val logger = coroutineContext.logger<ClockSkewInterceptor>()
+
+        val skew = currentSkew
+        if (skew != Duration.ZERO) {
+            logger.debug { "applying clock skew $skew to client" }
+            context.executionContext[HttpOperationContext.ClockSkew] = skew
+        }
```

Comment:
**Nit**: Extreme corner case but if the skew is ever non-zero and then is re-adjusted to zero, this will fail to update the execution context for existing requests.

Reply:
Handled by using `null` as the initial value for `currentSkew` and applying skew every time it's non-null. The side effect of this is if it's non-zero and then re-adjusted to zero, the zero skew will be applied to every request (it's not a problem, just a different code path)